core/arch/arm/include/kernel/misc.h

```diff
@@ -28,15 +28,82 @@
 #ifndef KERNEL_MISC_H
 #define KERNEL_MISC_H
 
+#include <mm/mobj.h>
 #include <types_ext.h>
 #include <arm.h>
 #include <kernel/thread.h>
 
+enum memparam_dir {
+	MEMPARAM_DIR_IN = 0,
+	MEMPARAM_DIR_OUT,
+	MEMPARAM_DIR_INOUT,
+};
+
 size_t get_core_pos(void);
 
 uint32_t read_mode_sp(int cpu_mode);
 uint32_t read_mode_lr(int cpu_mode);
 
+/**
+ * extract_pages_from_params() - extract list of pages from
+ * OPTEE_MSG_ATTR_FRAGMENT parameters
+ *
+ * @params:	pointer to parameters array
+ * @pages:	output array of page addresses
+ * @num_params: number of parameters in array
+ *
+ * return:
+ *   page count on success
+ *   <0 on error
+ */
+int extract_pages_from_params(struct optee_msg_param *params, paddr_t *pages,
+			      uint32_t num_pages);
+
+/**
+ * map_params_buffer() - map parameters buffer into OP-TEE VA space
+ * @pa_params - physical pointer to parameters
+ * @num_params - number of parameters
+ *
+ * return:
+ *  struct shmem_mapping of mapped buffer on success
+ *  NULL on error.
+ */
+struct mobj *map_params_buffer(paddr_t pa_params, uint32_t num_params);
+
+/**
+ * fill_memparam() - fill memory reference parameter for RPC call
+ * @param	- parameter to fill
+ * @mobj	- mobj describing the shared memory buffer
+ * @offset	- offset of the buffer
+ * @size	- size of the buffer
+ * @cookie	- NW cookie of the shared buffer
+ * @dir		- data direction
+ *
+ * Idea behind this function is that thread_rpc_alloc() can return
+ * either buffer from preallocated memory pool, of buffer constructed
+ * from supplicant's memory. In first case parameter will have type
+ * OPTEE_MSG_ATTR_TYPE_TMEM_* and OPTEE_MSG_ATTR_TYPE_RMEM_ in second case.
+ * This function will fill parameter structure with right type, depending on
+ * the passed mobj.
+ *
+ * return:
+ *	true on success, false on failure
+ */
+bool fill_memparam(struct optee_msg_param *param, struct mobj *mobj,
```

Comment:
Please use `msg_param` instead, the name of this function could then be `msg_param_fill_memparam()`. Perhaps "fill" should be replaced by "init".